File: nu/backend/models/MealOrder.js
class MealOrder {
  constructor({
    orderID,
    requestDate,
    status,

    pilgrimID,
    pilgrimName,

    mealID,

    mealName,
    mealName_ar,
    mealName_en,

    mealType,
    mealType_ar,
    mealType_en,

    campaignID,
    campaignName,
    campaignNumber,

    stars,
  }) {
    this.orderID = orderID;
    this.requestDate = requestDate;
    this.status = status;

    this.pilgrimID = pilgrimID;
    this.pilgrimName = pilgrimName;

    this.mealID = mealID;

    // القديم نخليه عشان الفرونت القديم ما ينكسر
    this.mealName = mealName;
    this.mealType = mealType;

    // الجديد
    this.mealName_ar = mealName_ar;
    this.mealName_en = mealName_en;

    this.mealType_ar = mealType_ar;
    this.mealType_en = mealType_en;

    this.campaignID = campaignID;
    this.campaignName = campaignName;
    this.campaignNumber = campaignNumber;

    this.stars = stars;
  }

  static fromRow(row) {
    const mealName =
      row.mealName ||
      row.mealName_en ||
      row.mealName_ar ||
      '';

    const mealNameAr =
      row.mealName_ar ||
      row.mealName ||
      row.mealName_en ||
      '';

    const mealNameEn =
      row.mealName_en ||
      row.mealName ||
      row.mealName_ar ||
      '';

    const mealType =
      row.mealType ||
      row.mealType_en ||
      row.mealType_ar ||
      '';

    const mealTypeAr =
      row.mealType_ar ||
      row.mealType ||
      row.mealType_en ||
      '';

    const mealTypeEn =
      row.mealType_en ||
      row.mealType ||
      row.mealType_ar ||
      '';

    return new MealOrder({
      orderID: row.orderID,
      requestDate: row.requestDate,
      status: row.status,

      pilgrimID: row.pilgrimID,
      pilgrimName: row.pilgrimName,

      mealID: row.mealID,

      mealName,
      mealName_ar: mealNameAr,
      mealName_en: mealNameEn,

      mealType,
      mealType_ar: mealTypeAr,
      mealType_en: mealTypeEn,

      campaignID: row.campaignID,
      campaignName: row.campaignName,
      campaignNumber: row.campaignNumber,

      stars: row.stars,
    });
  }
}

module.exports = MealOrder;

File: nu/lib/models/meal_order.dart
class MealOrder {
  final int orderID;
  final DateTime requestDate;
  final String status;
  final String pilgrimID;
  final String pilgrimName;
  final int mealID;

  // Old fields
  final String mealName;
  final String mealType;

  // New bilingual fields
  final String mealNameEn;
  final String mealNameAr;
  final String mealTypeEn;
  final String mealTypeAr;

  final String? campaignID;
  final String campaignName;
  final String campaignNumber;
  final bool isReviewed;
  final int? reviewRating;

  MealOrder({
    required this.orderID,
    required this.requestDate,
    required this.status,
    required this.pilgrimID,
    required this.pilgrimName,
    required this.mealID,
    required this.mealName,
    required this.mealType,
    this.mealNameEn = '',
    this.mealNameAr = '',
    this.mealTypeEn = '',
    this.mealTypeAr = '',
    required this.campaignID,
    required this.campaignName,
    required this.campaignNumber,
    required this.isReviewed,
    required this.reviewRating,
  });

  static int _toInt(dynamic value) {
    if (value == null) return 0;
    if (value is int) return value;
    if (value is num) return value.toInt();

    return int.tryParse(value.toString()) ?? 0;
  }

  static String _toStringValue(dynamic value) {
    if (value == null) return '';
    return value.toString();
  }

  factory MealOrder.fromJson(Map<String, dynamic> json) {
    final stars = json['stars'];

    final oldMealName = _toStringValue(json['mealName']);
    final oldMealType = _toStringValue(json['mealType']);

    final mealNameEn = _toStringValue(json['mealName_en']);
    final mealNameAr = _toStringValue(json['mealName_ar']);

    final mealTypeEn = _toStringValue(json['mealType_en']);
    final mealTypeAr = _toStringValue(json['mealType_ar']);

    return MealOrder(
      orderID: _toInt(json['orderID']),
      requestDate:
          DateTime.tryParse(json['requestDate']?.toString() ?? '') ??
          DateTime.now(),
      status: json['status']?.toString() ?? 'pending',
      pilgrimID: json['pilgrimID']?.toString() ?? '',
      pilgrimName: json['pilgrimName']?.toString() ?? 'Unknown Pilgrim',
      mealID: _toInt(json['mealID']),

      mealName: oldMealName.isNotEmpty
          ? oldMealName
          : mealNameEn.isNotEmpty
          ? mealNameEn
          : mealNameAr.isNotEmpty
          ? mealNameAr
          : 'Unknown Meal',

      mealType: oldMealType.isNotEmpty
          ? oldMealType
          : mealTypeEn.isNotEmpty
          ? mealTypeEn
          : mealTypeAr.isNotEmpty
          ? mealTypeAr
          : 'Meal',

      mealNameEn: mealNameEn.isNotEmpty ? mealNameEn : oldMealName,
      mealNameAr: mealNameAr.isNotEmpty ? mealNameAr : oldMealName,
      mealTypeEn: mealTypeEn.isNotEmpty ? mealTypeEn : oldMealType,
      mealTypeAr: mealTypeAr.isNotEmpty ? mealTypeAr : oldMealType,

      campaignID: json['campaignID']?.toString(),
      campaignName: json['campaignName']?.toString() ?? 'Unknown Campaign',
      campaignNumber: json['campaignNumber']?.toString() ?? '',
      reviewRating: stars == null ? null : _toInt(stars),
      isReviewed: stars != null,
    );
  }

  String localizedMealName(String languageCode) {
    final isArabic = languageCode.toLowerCase().startsWith('ar');

    if (isArabic && mealNameAr.trim().isNotEmpty) {
      return mealNameAr;
    }

    if (!isArabic && mealNameEn.trim().isNotEmpty) {
      return mealNameEn;
    }

    return mealName;
  }

  String localizedMealType(String languageCode) {
    final isArabic = languageCode.toLowerCase().startsWith('ar');

    if (isArabic && mealTypeAr.trim().isNotEmpty) {
      return mealTypeAr;
    }

    if (!isArabic && mealTypeEn.trim().isNotEmpty) {
      return mealTypeEn;
    }

    return mealType;
  }

  String get formattedRequestDate {
    final d = requestDate.day.toString().padLeft(2, '0');
    final m = requestDate.month.toString().padLeft(2, '0');
    final y = requestDate.year.toString();

    return '$d/$m/$y';
  }
}
